text_generation/causal_lm/cpp/continuous_batching/library/src/block_manager.hpp

```diff
@@ -141,16 +145,58 @@ class BlockManager {
         OPENVINO_ASSERT(m_block_table.erase(seq_id) == 1);
     }
 
-    bool can_append_slot(SequenceGroup::CPtr seq_group) {
-        // TODO: optimize this HEURISTIC
-        // it assumes that all sequences require new block, but maybe some of them
-        // don't share the same block
-        // let's count actual number of sequences, where last_block_id is the same
-        return seq_group->num_running_seqs() <= m_allocator.num_free_blocks();
+    void free_sequence_partially(size_t seq_id, size_t block_num) {
+        auto block_table = m_block_table[seq_id];
+
+        if (block_num == block_table.size())
+            return free_sequence(seq_id);
+
+        OPENVINO_ASSERT(block_table.size() >= block_num);
+        for (size_t idx = 0; idx < block_num; idx++) {
+            while (!block_table.back()->is_free()) {
+                m_allocator.free(block_table.back());
+            }
+        } 
+        m_block_table[seq_id].resize(m_block_table[seq_id].size() - block_num);
+
+        if (m_block_table.size() == 0) {
+            OPENVINO_ASSERT(m_block_table.erase(seq_id) == 1);
+        }
     }
 
-    std::map<size_t, std::list<size_t>> append_slot(SequenceGroup::CPtr seq_group) {
-        OPENVINO_ASSERT(can_append_slot(seq_group));
+    bool can_append_slots(SequenceGroup::CPtr seq_group) {
+        return required_blocks_count(seq_group) <= m_allocator.num_free_blocks();
+    }
+
+    size_t required_blocks_count(SequenceGroup::CPtr seq_group) {
+        std::vector<Sequence::CPtr> running_sequences = seq_group->get_running_sequences();
+        size_t blocks_count= 0; // totat number of needed blocks for sequence group
+
+        for (auto seq: running_sequences) {
+            auto seq_id = seq->get_id();
+            if (m_block_table.find(seq_id) == m_block_table.end()) {
+                // the block table is empty, so we need to allocate the number of blocks equal to number of logical blocks
+                blocks_count += seq_group->get_num_logical_blocks();
+                continue;
+            }
+            auto& block_table = m_block_table[seq_id];
+            size_t num_physical_blocks = block_table.size();
+            OPENVINO_ASSERT(num_physical_blocks > 0);
+            if (num_physical_blocks < seq_group->get_num_logical_blocks()) {
+                size_t needed_blocks_for_current_seq = seq_group->get_num_logical_blocks() - num_physical_blocks;
+
+                // If last block needed to be copied add 1 additional block to required blocks
+                KVCacheBlock::Ptr last_block = block_table.back();
+                if (needed_blocks_for_current_seq == 0 && last_block->copy_on_write())
+                    needed_blocks_for_current_seq+=1;
```

Comment:
if 2 sequences shares the same last block, we need to add only one new block, but here we will add 2

Reply:
Changed logic here as discussed. Please take a look.